extensions/amp-story/0.1/page-element.js

```diff
@@ -0,0 +1,366 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A map of elements to delay showing the page.  The key is a DOM query to find
+ * all elements that should wait for the specified event; the value is a factory
+ * method that will return a PageElement to define rendering and loading
+ * strategies.
+ *
+ * @const {!Object<string, !function(!Element, !AmpStoryPage): !PageElement>}
+ */
+const PAGE_ELEMENT_FACTORIES = {
+  'amp-audio, amp-video, .i-amphtml-story-background-audio':
+      (element, page) => new MediaElement(element, page),
+  'amp-img, amp-anim':
+      (element, page) => new ImageElement(element, page),
+  '.i-amphtml-video-interface':
+      (element, page) => new VideoInterfaceElement(element, page),
+};
+
+
+/**
+ * CSS class for an element on an amp-story-page.
+ * @const {string}
+ */
+const ELEMENT_CLASS_NAME = 'i-amphtml-story-page-element';
+
+
+/**
+ * CSS class for an element on an amp-story-page that indicates the element is
+ * loaded.
+ * @const {string}
+ */
+const ELEMENT_LOADED_CLASS_NAME = 'i-amphtml-story-page-element-loaded';
+
+
+/**
+ * CSS class for an element on an amp-story-page that indicates the element can
+ * be shown in the UI.
+ * @const {string}
+ */
+const ELEMENT_SHOW_CLASS_NAME = 'i-amphtml-story-page-element-shown';
+
+
+/**
+ * CSS class for an element on an amp-story-page that indicates the element has
+ * failed to load.
+ * @const {string}
+ */
+const ELEMENT_FAILED_CLASS_NAME = 'i-amphtml-story-page-element-failed';
+
+
+/**
+ * The minimum amount of a media item (by percentage) that must be loaded in
+ * order for that element to be considered "loaded".  Note that if the total
+ * size cannot be determined, this criteria is simply ignored.
+ */
+const MINIMUM_MEDIA_BUFFER_PERCENTAGE_FROM_BEGINNING = 0.25;
+
+
+/**
+ * The minimum amount of a media item (in seconds) that must be loaded in order
+ * for that element to be considered "loaded".
+ */
+const MINIMUM_MEDIA_BUFFER_SECONDS_FROM_BEGINNING = 3;
+
+
+
+export class PageElement {
+  /**
+   * @param {!Element} element The element on the page.
+   * @param {!AmpStoryPage} page The page that the element is on.
+   */
+  constructor(element, page) {
+    /** @protected @const {!Element} */
+    this.element = element;
+    this.element.classList.add(ELEMENT_CLASS_NAME);
+
+    /** @protected @const {!AmpStoryPage} */
+    this.page = page;
+
+    /** @public {boolean} */
+    this.isLoaded = false;
+
+    /** @public {boolean} */
+    this.canBeShown = false;
+
+    /** @public {boolean} */
+    this.failed = false;
```

Comment:
If these states are mutually exclusive, consider replacing with an enum. This will avoid bugs like `isLoaded == true && failed == true)`.

Reply:
The conditions are a bit more complicated than that.  `isLoaded` and `failed` are mutually exclusive, and `canBeShown` must be true if either `isLoaded` or `failed` are true.  The matrix is:

`isLoaded` | `canBeShown` | `failed` | allowed?
------------|----------------|----------|--
true | true | true | disallowed
true | true | false | allowed
true | false | true | disallowed
true | false | false | disallowed
false | true | true | allowed
false | true | false | allowed
false | false | true | disallowed
false | false | false | allowed

Filed #11534 to track this refactor.